### PR TITLE
KEYCLOAK-8976 Login after registration fails when other user was logg…

### PR DESCRIPTION
**Pull request by martin-kanis**
…ed in before

### CODE REVIEW

File: services/src/main/java/org/keycloak/protocol/AuthorizationEndpointBase.java [hunk 1]
```diff
@@ -188,16 +189,19 @@ protected AuthenticationSessionModel createAuthenticationSession(ClientModel cli
...
             if (userSession != null) {
-                String userSessionId = userSession.getId();
-                rootAuthSession = session.authenticationSessions().createRootAuthenticationSession(userSessionId, realm);
-                authSession = rootAuthSession.createAuthenticationSession(client);
-                logger.debugf("Sent request to authz endpoint. We don't have root authentication session with ID '%s' but we have userSession." +
-                        "Re-created root authentication session with same ID. Client is: %s . New authentication session tab ID: %s", userSessionId, client.getClientId(), authSession.getTabId());
+                if (AuthorizationEndpoint.Action.REGISTER.equals(action)) {
+                    authSession = createNewAuthenticationSession(manager, client);
+
+                    session.sessions().removeUserSession(realm, userSession);
```
Comment (mposolda) [inline services/src/main/java/org/keycloak/protocol/AuthorizationEndpointBase.java:195]: Few points to this:

1) It may be slightly better to call "AuthenticationManager.backchannelLogout" That one also cleanup cookies, logout all clients from the expired userSession etc.

2) I can think of the scenario like this:
- User is logged
- The registration URL is opened (for example after press "Register" link from the application)
- User then decides to go "Back" from the registration form. For example he pressed "Register" link by mistake in the previous step. But he won't be able to go back as session is already logged-out.

So my vote is instead to handle this in AuthenticationManager.attachSession (which is triggered after registration flow is successfully finished) rather than at the beginning of the flow.
- Now the session is already

**Review by mposolda — CHANGES_REQUESTED** [review]
@martin-kanis Thanks for the PR. My vote is to address this slightly differently. Hopefully it won't be big issue.

Also I vote to address this PR and https://github.com/keycloak/keycloak/pull/6977 together as they are quite related and IMO it makes sense to test them together - maybe use different commits for both JIRAs, but I would use same PR.

File: services/src/main/java/org/keycloak/authentication/AuthenticationProcessor.java [hunk 2]
```diff
@@ -1015,13 +1014,15 @@ public static ClientSessionContext attachSession(AuthenticationSessionModel auth
...
                         , remember, brokerSessionId, brokerUserId);
             } else {
-                // We have existing userSession even if it wasn't attached to authenticator. Could happen if SSO authentication was ignored (eg. prompt=login) and in some other cases.
-                // We need to handle case when different user was used
-                logger.debugf("No SSO login, but found existing userSession with ID '%s' after finished authentication.", userSession.getId());
                 if (!authSession.getAuthenticatedUser().equals(userSession.getUser())) {
-                    event.detail(Details.EXISTING_USER, userSession.getUser().getId());
-                    event.error(Errors.DIFFERENT_USER_AUTHENTICATED);
-                    throw new ErrorPageException(session, authSession, Response.Status.INTERNAL_SERVER_ERROR, Messages.DIFFERENT_USER_AUTHENTICATED, userSession.getUser().getUsername());
+                    AuthenticationManager.backchannelLogout(session, userSession, true);
```
Comment (martin-kanis) [inline services/src/main/java/org/keycloak/authentication/AuthenticationProcessor.java:1018]: @mposolda Thank you for the review and suggestions. I'm not sure about this particular change. I removed the original branch where an error was thrown. Maybe we need to somehow check from what "state" we ended up here where we have one active user session but different user in authSession. By "state" I mean for example we came here from registration form and we want to log in different user or the original user was disabled and we want to log in different user. 
The bottom line: I am not sure whether we can always log out original user and create new authSession without checking why this situation happened.

Comment (mposolda) [reply services/src/main/java/org/keycloak/authentication/AuthenticationProcessor.java:1018]: Well, my hope is, we can always logout original user and create new userSession. This can be for all flows like "register", "reset-credentials" or "login" and will allow to avoid some additional complexity.

BTV. There is discussion on keycloak-dev regarding this, so I suggest to continue discussion there.

**Comment by mposolda** [comment]
@martin-kanis Discussed with @stianst that we're closing this PR for now as we're going with different approach to address this JIRA.

**Comment by vincentpalita** [comment]
@mposolda do you have any update on this? I cannot find the Jira ticket linked to that issue.
Thank you!

**Comment by vincentpalita-whoz** [comment]
@mposolda Ping :)

**Comment by vincentpalita** [comment]
@mposolda ping :)

**Comment by Lowentwickler** [comment]
@mposolda Hello Marek. The problem is still there. Any plans on fixing?
